src/apps/schedule/crud/user_device_events.py

```diff
@@ -0,0 +1,43 @@
+__all__ = ["UserDeviceEventsCRUD"]
+
+import datetime
+import uuid
+
+from sqlalchemy.dialects.postgresql import Insert
+from sqlalchemy.dialects.postgresql import insert as pg_insert
+
+from apps.schedule.db.schemas import UserDeviceEventsSchema
+from infrastructure.database import BaseCRUD
+
+
+class UserDeviceEventsCRUD(BaseCRUD[UserDeviceEventsSchema]):
+    async def record_event_versions(
+        self,
+        device_id: uuid.UUID,
+        event_versions: list[tuple[uuid.UUID, str]],
+    ) -> list[UserDeviceEventsSchema]:
+        values = [
+            dict(
+                device_id=device_id,
+                event_id=event_id,
+                version=event_version,
+            )
+            for event_id, event_version in event_versions
+        ]
+
+        upsert: Insert = pg_insert(UserDeviceEventsSchema)
+        upsert = upsert.values(values)
+        upsert = upsert.on_conflict_do_update(
+            constraint=UserDeviceEventsSchema.unique_constraint,
+            set_={
+                "version": upsert.excluded.version,
+                "updated_at": datetime.datetime.now(datetime.UTC).replace(tzinfo=None),
+            },
+        )
+        upsert = upsert.returning(UserDeviceEventsSchema)
+        result = await self._execute(upsert)
+
+        rows = result.mappings().all()
+        model = [UserDeviceEventsSchema(**row) for row in rows]
+
+        return model
```

Comment:
```suggestion
        rows = result.scalars().all()

        return rows
```

We are returning UserDeviceEventsSchema, so better to use scalars().all(), it returns a list of UserDeviceEventsSchema objects

Reply:
I'm not sure why this doesn't work, but when I try this I just get a list of `bool`s

<img width="616" alt="image" src="https://github.com/user-attachments/assets/30b3c6ba-b716-4671-8274-811524f04aaa" />

<img width="618" alt="image" src="https://github.com/user-attachments/assets/1eff0eeb-0862-4b92-9eb4-3607f9210748" />

I assume these correspond to which rows were updated because they're `False` when the rows are inserted